File: admin/src/shared/api/admin-prompt-packs.ts
// 作者: 杨永的Agent
// 日期: 2026-05-21
// 修改功能: **`PromptPackSummary`/`PromptPackDetail.title`** · **`resolvePromptPackDisplayTitle`**
// 作者: 杨永的Agent
// 日期: 2026-05-20
// 修改功能: **`CreatePromptPackBody.sourcePromptPackId`**（从模板复制 · FE_HANDOFF 新建草稿）
// 作者: 杨永的Agent
// 日期: 2026-05-20
// 修改功能: fork / versions / rollback；PATCH 可选 If-Match（rowVersion）；详情 bodyMarkdown / updatedAt
// 作者: 杨永的Agent
// 日期: 2026-05-19
// 修改功能: PATCH 支持 **`variableSchema`**（与 **`messages`** 至少其一）；**`PromptPackDetail.variableSchema`**（FE_HANDOFF AC-09f）
// 作者: 杨永的Agent
// 日期: 2026-05-17
// 修改功能: POST 创建 DRAFT、`…/publish`、列表 Query；保留 GET 详情与 PATCH messages（FE_HANDOFF 2026-05-15）
import { HTTPError } from 'ky'

import { AppError } from '@/shared/api/errors'
import { httpClient, toAppError } from '@/shared/api/http-client'

const PREFIX = 'v1/admin/prompt-packs'

async function normalizeKyError(error: unknown): Promise<never> {
  if (error instanceof HTTPError) {
    const cloned = error.response.clone()
    const body: unknown = await cloned.json().catch(() => null)
    if (
      body !== null &&
      typeof body === 'object' &&
      'detail' in body &&
      typeof (body as { detail: unknown }).detail === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { detail: string }).detail, {
        status: error.response.status,
        code,
      })
    }
    if (
      body !== null &&
      typeof body === 'object' &&
      'message' in body &&
      typeof (body as { message: unknown }).message === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { message: string }).message, {
        status: error.response.status,
        code,
      })
    }
  }
  throw await toAppError(error)
}

export interface PromptPackSummary {
  promptPackId: string
  promptPackType: string
  scenarioId: string | null
  /** 运营展示名（服务端解析 · 只读） */
  title?: string | null
  promptPackVersion: string | null
  lifecycle: string | null
  etag: string | null
}

export interface PromptPackListResponse {
  items: PromptPackSummary[]
}

export interface ResolvedPromptBinding {
  scenarioId?: string | null
  sessionId?: string | null
  systemPromptPackId?: string | null
  systemPromptPackVersion?: string | null
  safetyPromptPackId?: string | null
  safetyPromptPackVersion?: string | null
  tradingPromptPackId?: string | null
  tradingPromptPackVersion?: string | null
  runtimeClarifyPromptPackId?: string | null
  runtimeClarifyPromptPackVersion?: string | null
  runtimeOutputContractPromptPackId?: string | null
  runtimeOutputContractPromptPackVersion?: string | null
  fewShotDigest?: string | null
  placeholderDenylistRevision?: string | null
  safetyPhraseBlocklistRevision?: string | null
}

export interface PromptPackDetail {
  promptPackId: string
  promptPackType: string
  scenarioId: string | null
  /** 运营展示名（服务端解析 · 只读） */
  title?: string | null
  promptPackVersion: string
  lifecycle: string
  etag: string | null
  rowVersion?: number | null
  updatedAt?: string | null
  placeholderDenylistRevision?: string | null
  safetyPhraseBlocklistRevision?: string | null
  messages: Record<string, unknown>[]
  /** 编辑器正文（与 messages 中 system content 同源） */
  bodyMarkdown?: string | null
  resolvedPromptBinding: ResolvedPromptBinding
  /** 非空 JSON 对象时：**`messages` 内 `{{slug}}`** 须在 **本平台白名单 ∪ 本题键集合** · `nullable`（见服务端校验） */
  variableSchema?: Record<string, unknown> | null
}

export interface PromptPackVersionEntry {
  promptPackVersion: string
  publishedAt: string
  lifecycle?: string | null
  event?: string | null
  actor?: string | null
  summary?: string | null
}

export interface PromptPackVersionHistoryResponse {
  items: PromptPackVersionEntry[]
}

export interface PromptPackForkBody {
  promptPackId?: string
}

export interface PromptPackRollbackBody {
  promptPackVersion: string
}

export interface PatchPromptPackOptions {
  /** PM-C03：传 **`rowVersion`**（数字或已带引号的 ETag 字符串） */
  ifMatch?: string | number | null
}

/** PATCH：**`messages` / `variableSchema`** 可选；服务端要求 **至少传其一**（本模块在发包前兜底校验）。 */
export interface PromptPackPatchBody {
  messages?: Record<string, unknown>[]
  variableSchema?: Record<string, unknown> | null
}

/** @deprecated 请改用 **`PromptPackPatchBody`**；保留别名供旧 import。 */
export type PromptPackMessagesPatchBody = PromptPackPatchBody

/** 列表查询，与控制台筛选对齐；服务端未识别之键会自动忽略或由网关校验 */
export interface ListPromptPacksQuery {
  promptPackType?: string
  scenarioId?: string
  lifecycle?: string
}

export interface CreatePromptPackBody {
  promptPackType: string
  scenarioId?: string
  promptPackId?: string
  /** 从已有包复制 messages + variableSchema 到新 DRAFT */
  sourcePromptPackId?: string
}

export interface PromptPackPublishResult {
  promptPackId: string
  promptPackVersion: string
  lifecycle: string
}

/** 列表/编辑器展示名：优先 **`title`**，fallback scenarioId → promptPackId */
export function resolvePromptPackDisplayTitle(
  row: Pick<PromptPackSummary, 'title' | 'scenarioId' | 'promptPackId'>,
): string {
  const title = (row.title ?? '').trim()
  if (title) return title
  const sid = (row.scenarioId ?? '').trim()
  return sid || row.promptPackId
}

function withListQuery(prefix: string, query?: ListPromptPacksQuery): string {
  if (!query) return prefix
  const params = new URLSearchParams()
  const t = query.promptPackType?.trim()
  const s = query.scenarioId?.trim()
  const l = query.lifecycle?.trim()
  if (t) params.set('promptPackType', t)
  if (s) params.set('scenarioId', s)
  if (l) params.set('lifecycle', l)
  const qs = params.toString()
  return qs ? `${prefix}?${qs}` : prefix
}

export async function listPromptPacks(query?: ListPromptPacksQuery): Promise<PromptPackListResponse> {
  try {
    return await httpClient.get(withListQuery(PREFIX, query)).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getPromptPack(promptPackId: string): Promise<PromptPackDetail> {
  try {
    return await httpClient.get(`${PREFIX}/${encodeURIComponent(promptPackId)}`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

function ifMatchHeaderValue(ifMatch: string | number): string {
  const raw = String(ifMatch).trim()
  if (!raw) return raw
  return raw.startsWith('"') ? raw : `"${raw}"`
}

export async function patchPromptPack(
  promptPackId: string,
  body: PromptPackPatchBody,
  options?: PatchPromptPackOptions,
): Promise<PromptPackDetail> {
  const hasMessages = body.messages !== undefined
  const hasSchema = body.variableSchema !== undefined
  if (!hasMessages && !hasSchema) {
    throw new AppError('messages 与 variableSchema 至少须在 PATCH Body 中带其一。', {
      status: 400,
    })
  }
  const headers: Record<string, string> = {}
  if (options?.ifMatch != null && String(options.ifMatch).trim() !== '') {
    headers['If-Match'] = ifMatchHeaderValue(options.ifMatch)
  }
  try {
    return await httpClient
      .patch(`${PREFIX}/${encodeURIComponent(promptPackId)}`, { json: body, headers })
      .json<PromptPackDetail>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function forkPromptPack(
  sourcePromptPackId: string,
  body?: PromptPackForkBody,
): Promise<PromptPackDetail> {
  try {
    const payload: Record<string, string> = {}
    const pid = body?.promptPackId?.trim()
    if (pid) payload.promptPackId = pid
    return await httpClient
      .post(`${PREFIX}/${encodeURIComponent(sourcePromptPackId)}/fork`, {
        json: Object.keys(payload).length ? payload : {},
      })
      .json<PromptPackDetail>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function listPromptPackVersions(
  promptPackId: string,
): Promise<PromptPackVersionHistoryResponse> {
  try {
    return await httpClient
      .get(`${PREFIX}/${encodeURIComponent(promptPackId)}/versions`)
      .json<PromptPackVersionHistoryResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function rollbackPromptPack(
  promptPackId: string,
  body: PromptPackRollbackBody,
): Promise<PromptPackPublishResult> {
  try {
    return await httpClient
      .post(`${PREFIX}/${encodeURIComponent(promptPackId)}/rollback`, { json: body })
      .json<PromptPackPublishResult>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** PATCH 别名：与 **`patchPromptPack`** 等价。 */
export async function patchPromptPackMessages(
  promptPackId: string,
  body: PromptPackMessagesPatchBody,
): Promise<PromptPackDetail> {
  return patchPromptPack(promptPackId, body)
}

export async function createPromptPack(body: CreatePromptPackBody): Promise<PromptPackDetail> {
  try {
    const payload: Record<string, string> = {
      promptPackType: body.promptPackType.trim(),
    }
    const sid = body.scenarioId?.trim()
    const pid = body.promptPackId?.trim()
    const src = body.sourcePromptPackId?.trim()
    if (sid) payload.scenarioId = sid
    if (pid) payload.promptPackId = pid
    if (src) payload.sourcePromptPackId = src
    return await httpClient.post(PREFIX, { json: payload }).json<PromptPackDetail>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function publishPromptPack(promptPackId: string): Promise<PromptPackPublishResult> {
  try {
    return await httpClient
      .post(`${PREFIX}/${encodeURIComponent(promptPackId)}/publish`)
      .json<PromptPackPublishResult>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

File: server/chainup_agent/api/routers/admin_prompt_packs.py
"""Admin — Prompt packs Phase1（列表/详情/草稿/发布/fork/versions/rollback · messages PATCH）。"""

from __future__ import annotations

from chainup_agent.api.deps import DbSession, require_admin_console_bearer
from chainup_agent.api.schemas.prompt_management import (
    PromptPackCreateBody,
    PromptPackDetailOut,
    PromptPackForkBody,
    PromptPackListOut,
    PromptPackMessagesPatchBody,
    PromptPackSummaryOut,
    PromptPublishResultOut,
    PromptRollbackBody,
    PromptVersionEntryOut,
    PromptVersionHistoryOut,
)
from chainup_agent.application.admin_prompt_packs import (
    create_draft_prompt_pack,
    fork_prompt_pack_from_source,
    get_prompt_pack,
    list_prompt_pack_versions,
    list_prompt_packs,
    pack_to_detail_dict,
    pack_to_summary_dict,
    patch_messages,
    publish_prompt_pack,
    rollback_prompt_pack,
)
from chainup_agent.core.errors import AppError
from fastapi import APIRouter, Depends, Header, Query, Response, status

router = APIRouter(
    prefix="/api/v1/admin/prompt-packs",
    tags=["Admin — Prompt packs"],
    dependencies=[Depends(require_admin_console_bearer)],
)


def _set_row_version_etag(response: Response, row_version: int) -> None:
    response.headers["ETag"] = f'"{row_version}"'


@router.get(
    "",
    response_model=PromptPackListOut,
    response_model_by_alias=True,
    summary="Prompt 包列表（Phase1；可选筛选）",
)
async def admin_list_prompt_packs(
    db: DbSession,
    prompt_pack_type: str | None = Query(None, alias="promptPackType"),
    scenario_id: str | None = Query(None, alias="scenarioId"),
    lifecycle: str | None = Query(None),
) -> PromptPackListOut:
    rows = await list_prompt_packs(
        db,
        prompt_pack_type=prompt_pack_type,
        scenario_id=scenario_id,
        lifecycle=lifecycle,
    )
    return PromptPackListOut(
        items=[PromptPackSummaryOut.model_validate(pack_to_summary_dict(r)) for r in rows],
    )


@router.post(
    "",
    response_model=PromptPackDetailOut,
    response_model_by_alias=True,
    status_code=status.HTTP_201_CREATED,
    summary="创建草稿包（空白或 sourcePromptPackId 模板复制）",
)
async def admin_create_prompt_pack_draft(
    body: PromptPackCreateBody,
    response: Response,
    db: DbSession,
) -> PromptPackDetailOut:
    row = await create_draft_prompt_pack(
        db,
        prompt_pack_type=body.prompt_pack_type,
        scenario_id=body.scenario_id,
        prompt_pack_id=body.prompt_pack_id,
        source_prompt_pack_id=body.source_prompt_pack_id,
    )
    await db.commit()
    await db.refresh(row)
    _set_row_version_etag(response, int(row.row_version))
    return PromptPackDetailOut.model_validate(pack_to_detail_dict(row))


@router.get(
    "/{prompt_pack_id}",
    response_model=PromptPackDetailOut,
    response_model_by_alias=True,
    summary="Prompt 包详情",
)
async def admin_get_prompt_pack(
    prompt_pack_id: str,
    response: Response,
    db: DbSession,
) -> PromptPackDetailOut:
    row = await get_prompt_pack(db, prompt_pack_id)
    if row is None:
        raise AppError(
            code="AGENT_PROMPT_PACK_NOT_FOUND",
            message="未知的 promptPackId。",
            status_code=404,
            details={"promptPackId": prompt_pack_id.strip()},
        )
    _set_row_version_etag(response, int(row.row_version))
    return PromptPackDetailOut.model_validate(pack_to_detail_dict(row))


@router.patch(
    "/{prompt_pack_id}",
    response_model=PromptPackDetailOut,
    response_model_by_alias=True,
    summary="更新包的消息正文（DRAFT / PUBLISHED；LOCKED / 下线禁止 · PM-C03 If-Match）",
)
async def admin_patch_prompt_pack_messages(
    prompt_pack_id: str,
    body: PromptPackMessagesPatchBody,
    response: Response,
    db: DbSession,
    if_match: str | None = Header(default=None, alias="If-Match"),
) -> PromptPackDetailOut:
    try:
        row = await patch_messages(
            db,
            prompt_pack_id=prompt_pack_id,
            messages=body.messages,
            variable_schema=body.variable_schema,
            if_match=if_match,
        )
    except AppError:
        await db.rollback()
        raise
    if row is None:
        raise AppError(
            code="AGENT_PROMPT_PACK_NOT_FOUND",
            message="未知的 promptPackId。",
            status_code=404,
            details={"promptPackId": prompt_pack_id.strip()},
        )
    await db.commit()
    await db.refresh(row)
    _set_row_version_etag(response, int(row.row_version))
    return PromptPackDetailOut.model_validate(pack_to_detail_dict(row))


@router.post(
    "/{prompt_pack_id}/fork",
    response_model=PromptPackDetailOut,
    response_model_by_alias=True,
    status_code=status.HTTP_201_CREATED,
    summary="从已有包快照复制为新草稿版本线（SYSTEM LOCKED → 新版本编辑路径）",
)
async def admin_fork_prompt_pack(
    prompt_pack_id: str,
    response: Response,
    db: DbSession,
    body: PromptPackForkBody | None = None,
) -> PromptPackDetailOut:
    new_id = body.prompt_pack_id if body is not None else None
    try:
        row = await fork_prompt_pack_from_source(
            db,
            source_prompt_pack_id=prompt_pack_id,
            new_prompt_pack_id=new_id,
        )
    except AppError:
        await db.rollback()
        raise
    await db.commit()
    await db.refresh(row)
    _set_row_version_etag(response, int(row.row_version))
    return PromptPackDetailOut.model_validate(pack_to_detail_dict(row))


@router.get(
    "/{prompt_pack_id}/versions",
    response_model=PromptVersionHistoryOut,
    response_model_by_alias=True,
    summary="版本履历（Publish / Rollback 快照）",
)
async def admin_get_prompt_pack_versions(
    prompt_pack_id: str,
    db: DbSession,
) -> PromptVersionHistoryOut:
    row = await get_prompt_pack(db, prompt_pack_id)
    if row is None:
        raise AppError(
            code="AGENT_PROMPT_PACK_NOT_FOUND",
            message="未知的 promptPackId。",
            status_code=404,
            details={"promptPackId": prompt_pack_id.strip()},
        )
    items = await list_prompt_pack_versions(db, prompt_pack_id=prompt_pack_id)
    return PromptVersionHistoryOut(
        items=[PromptVersionEntryOut.model_validate(x) for x in items],
    )


@router.post(
    "/{prompt_pack_id}/publish",
    response_model=PromptPublishResultOut,
    response_model_by_alias=True,
    summary="发布草稿（FR-PM07 · SYSTEM→LOCKED；场景族→PUBLISHED 并顶替同 scenario 旧版）",
)
async def admin_publish_prompt_pack(
    prompt_pack_id: str,
    db: DbSession,
) -> PromptPublishResultOut:
    try:
        row = await publish_prompt_pack(db, prompt_pack_id=prompt_pack_id)
    except AppError as exc:
        if exc.code == "PROMPT_SAFETY_VIOLATION":
            await db.commit()
        else:
            await db.rollback()
        raise
    if row is None:
        raise AppError(
            code="AGENT_PROMPT_PACK_NOT_FOUND",
            message="未知的 promptPackId。",
            status_code=404,
            details={"promptPackId": prompt_pack_id.strip()},
        )
    await db.commit()
    await db.refresh(row)
    return PromptPublishResultOut(
        promptPackId=row.prompt_pack_id,
        promptPackVersion=row.prompt_pack_version,
        lifecycle=row.lifecycle,
    )


@router.post(
    "/{prompt_pack_id}/rollback",
    response_model=PromptPublishResultOut,
    response_model_by_alias=True,
    summary="回滚生效指针至历史版本快照（FR-PM · 指针回指）",
)
async def admin_rollback_prompt_pack(
    prompt_pack_id: str,
    body: PromptRollbackBody,
    db: DbSession,
) -> PromptPublishResultOut:
    try:
        row = await rollback_prompt_pack(
            db,
            prompt_pack_id=prompt_pack_id,
            target_version=body.prompt_pack_version,
        )
    except AppError:
        await db.rollback()
        raise
    if row is None:
        raise AppError(
            code="AGENT_PROMPT_PACK_NOT_FOUND",
            message="未知的 promptPackId。",
            status_code=404,
            details={"promptPackId": prompt_pack_id.strip()},
        )
    await db.commit()
    await db.refresh(row)
    return PromptPublishResultOut(
        promptPackId=row.prompt_pack_id,
        promptPackVersion=row.prompt_pack_version,
        lifecycle=row.lifecycle,
    )
